pkl-core/src/main/java/org/pkl/core/util/RrbTree.java

```diff
@@ -0,0 +1,2257 @@
+    @Override
+    public Node<T> endChild(boolean leftMost) {
+      return nodes[leftMost ? 0 : nodes.length - 1];
+    }
+
+    /** Adds a node as the first/leftmost or last/rightmost child */
+    @Override
+    public Node<T> addEndChild(boolean leftMost, Node<T> shorter) {
+      return insertInRelaxedAt(cumulativeSizes, nodes, shorter, leftMost ? 0 : nodes.length);
+    }
+
+    /** Adds kids as leftmost or rightmost of current children */
+    @Override
+    public Node<T> addEndChildren(boolean leftMost, Node<T>[] newKids) {
+      //            if (!thisNodeHasRelaxedCapacity(newKids.length)) {
+      //                throw new IllegalStateException("Can't add enough kids");
+      //            }
+      //            if (nodes[0].height() != newKids[0].height()) {
+      //                throw new IllegalStateException("Kids not same height");
+      //            }
+      @SuppressWarnings("unchecked")
+      Node<T>[] res = spliceIntoArrayAt(newKids, nodes, leftMost ? 0 : nodes.length, Node.class);
+      // TODO: Figure out which side we inserted on and do the math to adjust counts instead
+      // of looking them up.
+      return new Relaxed<>(makeSizeArray(res), res);
+    }
+
+    @Override
+    public int height() {
+      return nodes[0].height() + 1;
+    }
+
+    @Override
+    public int size() {
+      return cumulativeSizes[cumulativeSizes.length - 1];
+    }
+
+    /**
+     * Converts the index of an item into the index of the sub-node containing that item.
+     *
+     * @param treeIndex The index of the item in the tree
+     * @return The index of the immediate child of this node that the desired node resides in.
+     */
+    private int subNodeIndex(int treeIndex) {
+      // For radix=4 this is actually faster, or at least as fast...
+      //            for (int i = 0; i < cumulativeSizes.length; i++) {
+      //                if (treeIndex < cumulativeSizes[i]) {
+      //                    return i;
+      //                }
+      //            }
+      //            if (treeIndex == size()) {
+      //                return cumulativeSizes.length - 1;
+      //            }
+
+      // treeSize = cumulativeSizes[cumulativeSizes.length - 1]
+      // range of sub-node indices: 0 to cumulativeSizes.length - 1
+      // range of tree indices: 0 to treeSize
+      // liner interpolation:
+      //     treeIndex / treeSize ~= subNodeIndex / cumulativeSizes.length
+      // solve for endIdxSlot
+      //     cumulativeSizes.length * (treeIndex / treeSize) =  subNodeIndex
+      // Put division last
+      //     subNodeIndex = cumulativeSizes.length * treeIndex / treeSize
+      //
+      // Now guess the sub-node index (quickly).
+
+      var guessLong = ((long) cumulativeSizes.length * treeIndex) / size();
+      if (guessLong >= cumulativeSizes.length) {
+        // Guessed beyond end length - returning last item.
+        return cumulativeSizes.length - 1;
+      }
+      var guess = (int) guessLong;
```

Comment:
This is the actual bugfix: do math with `long` for the bounds check.

Original code: https://github.com/GlenKPeterson/Paguro/blob/963e8ed91833656236489fd5a92b20b54f47b4b9/src/main/java/org/organicdesign/fp/collections/RrbTree.java#L1477-L1481